chainntnfs/txconfnotifier.go

```diff
@@ -0,0 +1,239 @@
+
+// TxConfNotifier is used to register transaction confirmation notifications and
+// dispatch them as the transactions confirm. A client can request to be
+// notified when a particular transaction has sufficient on-chain confirmations
+// (or be notified immediately if the tx already does), and the TxConfNotifier
+// will watch changes to the blockchain in order to satisfy these requests.
+type TxConfNotifier struct {
+	// currentHeight is the height of the tracked blockchain. It is used to
+	// determine the number of confirmations a tx has and ensure blocks are
+	// connected and disconnected in order.
+	currentHeight uint32
+
+	// reorgSafetyLimit is the chain depth beyond which it is assumed a block
+	// will not be reorganized out of the chain. This is used to determine when
+	// to prune old confirmation requests so that reorgs are handled correctly.
+	// The coinbase maturity period is a reasonable value to use.
+	reorgSafetyLimit uint32
+
+	// reorgDepth is the depth of a chain organization that this system is being
+	// informed of. This is incremented as long as a sequence of blocks are
+	// disconnected without being interrupted by a new block.
+	reorgDepth uint32
+
+	// confNotifications is an index of notification requests by transaction
+	// hash.
+	confNotifications map[chainhash.Hash][]*ConfNtfn
+
+	// confTxsByInitialHeight is an index of watched transactions by the height
+	// that they are included at in the blockchain. This is tracked so that
+	// incorrect notifications are not sent if a transaction is reorganized out
+	// of the chain and so that negative confirmations can be recognized.
+	confTxsByInitialHeight map[uint32][]*chainhash.Hash
+
+	// ntfnsByConfirmHeight is an index of notification requests by the height
+	// at which the transaction will have sufficient confirmations.
+	ntfnsByConfirmHeight map[uint32][]*ConfNtfn
+}
+
+// NewTxConfNotifier creates a TxConfNotifier. The current height of the
+// blockchain is accepted as a parameter.
+func NewTxConfNotifier(startHeight uint32, reorgSafetyLimit uint32) *TxConfNotifier {
+	return &TxConfNotifier{
+		currentHeight:          startHeight,
+		reorgSafetyLimit:       reorgSafetyLimit,
+		confNotifications:      make(map[chainhash.Hash][]*ConfNtfn),
+		confTxsByInitialHeight: make(map[uint32][]*chainhash.Hash),
+		ntfnsByConfirmHeight:   make(map[uint32][]*ConfNtfn),
+	}
+}
+
+// Register handles a new notification request. The client will be notified when
+// the transaction gets a sufficient number of confirmations on the blockchain.
+// If the transaction has already been included in a block on the chain, the
+// confirmation details must be given as the txConf argument, otherwise it
+// should be nil. If the transaction already has the sufficient number of
+// confirmations, this dispatches the notification immediately.
+func (tcn *TxConfNotifier) Register(ntfn *ConfNtfn, txConf *TxConfirmation) {
+	if txConf == nil || txConf.BlockHeight > tcn.currentHeight {
```

Comment:
maybe add comment that this means the tx is unconfirmed

Reply:
Done.